### PR TITLE
[WIP DRAFT PR]: Cross chain token

### PR DESCRIPTION
**Pull request by Aniket965**
**Please DO NOT MERGE 🚨🚨**

### CODE REVIEW

**Review by arthcp — CHANGES_REQUESTED** [review]

File: src/CrossChainToken/CrossChainToken.sol [hunk 1]
```diff
@@ -0,0 +1,218 @@
+pragma solidity 0.8.13;
+
+import "solmate/tokens/ERC20.sol";
+import "solmate/auth/Owned.sol";
+import "../Gauge.sol";
+import "./CrossChainConnector.sol";
+
+
+contract CrossChainToken is ERC20, Gauge, Owned, ISocketReceiver {
+
```
[the remark this answers is not in the document]
Comment (Aniket965) [reply src/CrossChainToken/CrossChainToken.sol:10]: it is coz we need to transfer ownership in contract itself  to owner

File: src/CrossChainToken/CrossChainToken.sol [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        emit LimitParamsUpdated(updates_);
+    }
+
+    function bridgeToChain(
+        address receiver_,
+        uint256 sendingAmount_,
+        uint256 msgGasLimit_,
+        address destinationConnector_
+    ) external payable {
```
[the remark this answers is not in the document]
Comment (Aniket965) [reply src/CrossChainToken/CrossChainToken.sol:101]: okies,

File: src/CrossChainToken/CrossChainToken.sol [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+            (address, uint256, bytes32)
+        );
+
+        (uint256 consumedAmount, uint256 pendingAmount) = _consumePartLimit(
+            mintAmount,
+            _receivingLimitParams[msg.sender]
+        );
+
+        if (pendingAmount > 0) {
+            pendingMints[msg.sender][receiver][identifier] += pendingAmount;
```
[the remark this answers is not in the document]
Comment (Aniket965) [reply src/CrossChainToken/CrossChainToken.sol:168]: ✅

File: src/CrossChainToken/DeployCrossChainTokenStack.sol [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        );
+
+        CrossChainToken cct = new CrossChainToken{salt:tokenSalt}(
+            data.tokenName,
+            data.tokenSymbol,
+            data.tokenDecimals,
+            data.owner,
+            address(this),
+            initialSupply
+        );
```
[the remark this answers is not in the document]
Comment (Aniket965) [reply src/CrossChainToken/DeployCrossChainTokenStack.sol:91]: yes they can't deploy it again with the same details

File: src/CrossChainToken/DeployCrossChainTokenStack.sol [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+            unchecked {
+                i++;
+            }
+        }
+
+        cct.updateLimitParams(updates);
+        cct.transferOwnership(data.owner);
+    }
+
+    function deployMultiChain(DeployToChains calldata data) external payable {
```
[the remark this answers is not in the document]
Comment (Aniket965) [reply src/CrossChainToken/DeployCrossChainTokenStack.sol:147]: i don't know bros why will someone pass excess here

File: src/CrossChainToken/DeployCrossChainTokenStack.sol [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+            );
+
+
+        bytes32 tokenSalt = keccak256(
+                abi.encodePacked(
+                    data.tokenName,
+                    data.tokenSymbol,
+                    data.tokenDecimals,
+                    data.owner,
+                    msg.sender,
```
[the remark this answers is not in the document]
Comment (Aniket965) [reply src/CrossChainToken/DeployCrossChainTokenStack.sol:79]: good catch I recently added this msg.sender will handle it

File: test/CrossChainToken.t.sol [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+            _msgGasLimit,
+            _wrongConnector
+        );
+    }
+
+    function testBridgeToChainMe() external {
+        _setLimits();
+        bytes32 messageIdMsg = bytes32(uint256(2));
+        bytes32 messageId = bytes32(uint256(1));
+
```
[the remark this answers is not in the document]
Comment (Aniket965) [reply test/CrossChainToken.t.sol:129]: okies

File: src/CrossChainToken/DeployCrossChainTokenStack.sol [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+struct DeployToChains {
+    uint256[] gasLimits;
+    uint256[] values;
+    uint256 initialSupply;
+    DeploymentInfo data;
+}
+
+contract DeployCrossChainTokenStack is IPlug, Owned {
+    event TokenDeployed(address token, address owner);
+    event ControllerDeployed(address controller, address token);
```
[the remark this answers is not in the document]
Comment (Aniket965) [reply src/CrossChainToken/DeployCrossChainTokenStack.sol]: removed

File: src/CrossChainToken/DeployCrossChainTokenStack.sol [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+                        chainSlug < data.chains[i] ? data.chains[i] : chainSlug,
+                        "connector"
+                    )
+                );
+               CrossChainConnector connector = new CrossChainConnector{
+                    salt: connectorSalt
+                }(address(cct), socket, data.chains[i]);
+
+                connector.connect(address(connector), switchboard, switchboard);
+                emit ConnectorDeployedAndConnected(
```
[the remark this answers is not in the document]
Comment (Aniket965) [reply src/CrossChainToken/DeployCrossChainTokenStack.sol:117]: fixed

File: src/CrossChainToken/DeployCrossChainTokenStack.sol [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+                        data.tokenName,
+                        data.tokenSymbol,
+                        data.tokenDecimals,
+                        data.owner,
+                        msg.sender,
+                        chainSlug < data.chains[i] ? chainSlug : data.chains[i],
+                        chainSlug < data.chains[i] ? data.chains[i] : chainSlug,
+                        "connector"
+                    )
+                );
```
[the remark this answers is not in the document]
Comment (Aniket965) [reply src/CrossChainToken/DeployCrossChainTokenStack.sol]: we only let 1 deployment per sender

[the remark this answers is not in the document]
Comment (Aniket965) [reply src/CrossChainToken/DeployCrossChainTokenStack.sol]: with same details

File: src/CrossChainToken/DeployCrossChainTokenStack.sol [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+               CrossChainConnector connector = new CrossChainConnector{
+                    salt: connectorSalt
+                }(address(cct), socket, data.chains[i]);
+
+                connector.connect(address(connector), switchboard, switchboard);
```
[the remark this answers is not in the document]
Comment (Aniket965) [reply src/CrossChainToken/DeployCrossChainTokenStack.sol:116]: dones

File: src/CrossChainToken/DeployCrossChainTokenStack.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (Aniket965) [reply src/CrossChainToken/DeployCrossChainTokenStack.sol:149]: let me know if you still want me to add, i will sum(data.values) and send rest of msg.value back

[the remark this answers is not in the document]
Comment (arthcp) [reply src/CrossChainToken/DeployCrossChainTokenStack.sol:149]: mostly by mistake

**Comment by arthcp** [comment]
stale

**Pull request closed without merging**